d2/src/main/java/com/linkedin/d2/balancer/clients/TrackerClient.java

```diff
@@ -13,450 +13,40 @@
-         * update the latency if it's successful, or update the error count if it's not successful.
-         * In this way, D2 still monitors the responsiveness of a server without the interference from the client
-         * side events, and error counting still works as before.
-         */
-        _callCompletion.record();
-        Observer observer = new Observer()
-        {
-          @Override
-          public void onDataAvailable(ByteString data)
-          {
-          }
-
-          @Override
-          public void onDone()
-          {
-            _callCompletion.endCall();
-          }
-
-          @Override
-          public void onError(Throwable e)
-          {
-            handleError(_callCompletion, e);
-          }
-        };
-        entityStream.addObserver(observer);
-      }
-
-      _wrappedCallback.onResponse(response);
-    }
-  }
-
-  private void handleError(CallCompletion callCompletion, Throwable throwable)
-  {
-    if (isServerError(throwable))
-    {
-      callCompletion.endCallWithError(ErrorType.SERVER_ERROR);
-    }
-    else if (throwable instanceof RemoteInvocationException)
-    {
-      Throwable originalThrowable = LoadBalancerUtil.findOriginalThrowable(throwable);
-      if (originalThrowable instanceof ConnectException)
-      {
-        callCompletion.endCallWithError(ErrorType.CONNECT_EXCEPTION);
-      }
-      else if (originalThrowable instanceof ClosedChannelException)
-      {
-        callCompletion.endCallWithError(ErrorType.CLOSED_CHANNEL_EXCEPTION);
-      }
-      else if (originalThrowable instanceof TimeoutException)
-      {
-        callCompletion.endCallWithError(ErrorType.TIMEOUT_EXCEPTION);
-      }
-      else
-      {
-        callCompletion.endCallWithError(ErrorType.REMOTE_INVOCATION_EXCEPTION);
-      }
-    }
-    else
-    {
-      callCompletion.endCallWithError();
-    }
-  }
 
   /**
-   * Returns true if the given throwable indicates a server-side error.
+   * @return CallStats tracked in the latest interval.
    */
-  private boolean isServerError(Throwable throwable)
-  {
-    if (throwable instanceof RestException)
-    {
-      RestException restException = (RestException) throwable;
-      if (restException.getResponse() != null)
-      {
-        return matchErrorStatus(restException.getResponse().getStatus());
-      }
-    }
-    else if (throwable instanceof StreamException)
-    {
-      StreamException streamException = (StreamException) throwable;
-      if (streamException.getResponse() != null)
-      {
-        return matchErrorStatus(streamException.getResponse().getStatus());
-      }
-    }
-    // default to false
-    return false;
-  }
+  CallTracker.CallStats getLatestCallStats();
 
-  private boolean matchErrorStatus(int status) {
-    return _errorStatusPattern.matcher(Integer.toString(status)).matches();
-  }
+  /**
+   * @return {@link PartitionData} map.
+   */
+  Map<Integer, PartitionData> getPartitionDataMap();
```

Comment:
Github does not allow replies to the top comment so I will include it here.

"Looking forward to further refactoring -- degraderControl, DegraderTrackerClientUpdater etc."
Actually, no further refactoring will be needed since we are getting rid of those classes in relative strategy. They will only be used by degrader.

"I think PartitionData is not necessary -- can we remove this class and just define a partitionWeightMap?"
Based on the original comment in PartitionData class it seems like it exists because partition could be extended to contain more than just the weight, which I think is valid. Are you strongly for removing the class? I'm okay with keeping it as is.

Reply:
We do have partitionProperties so the so-called partitionData simply complicated the code. A partitionWeight map is sufficient for the use case.

Reply:
PartitionProperties is for cluster-level, but PartitionData is URI specific. I can remove PartitionData if you feel strongly about it but PartitionProperties is not really a replacement for PartitionData.

Reply:
We probably won't expand uri level partition data. But let's keep it just in case.